infrastructures/kubernetes/src/main/java/org/eclipse/che/workspace/infrastructure/kubernetes/server/external/ExternalServerPathDemangler.java

```diff
@@ -0,0 +1,81 @@
+/*
+ * Copyright (c) 2012-2018 Red Hat, Inc.
+ * This program and the accompanying materials are made
+ * available under the terms of the Eclipse Public License 2.0
+ * which is available at https://www.eclipse.org/legal/epl-2.0/
+ *
+ * SPDX-License-Identifier: EPL-2.0
+ *
+ * Contributors:
+ *   Red Hat, Inc. - initial API and implementation
+ */
+package org.eclipse.che.workspace.infrastructure.kubernetes.server.external;
+
+import io.fabric8.kubernetes.api.model.HasMetadata;
+import java.util.regex.Matcher;
+import java.util.regex.Pattern;
+import javax.inject.Inject;
+import javax.inject.Named;
+import org.eclipse.che.commons.annotation.Nullable;
+
+public class ExternalServerPathDemangler {
+  private final Pattern pathTransformInverse;
+
+  @Inject
+  public ExternalServerPathDemangler(
+      @Nullable @Named("che.infra.kubernetes.ingress.path_transform") String pathTransformFmt) {
+    this.pathTransformInverse = extractPathFromFmt(pathTransformFmt);
+  }
+
+  private static Pattern extractPathFromFmt(String pathTransformFmt) {
+    int refIdx = pathTransformFmt.indexOf("%s");
+    String matchPath = "(.*)";
+
+    String transformed;
+    if (refIdx < 0) {
+      transformed = Pattern.quote(pathTransformFmt);
+    } else {
+      if (refIdx == 0) {
+        if (pathTransformFmt.length() > 2) {
+          transformed = matchPath + Pattern.quote(pathTransformFmt.substring(2));
+        } else {
+          transformed = matchPath;
+        }
+      } else {
+        String prefix = Pattern.quote(pathTransformFmt.substring(0, refIdx));
+        String suffix =
+            refIdx < pathTransformFmt.length() - 2
+                ? Pattern.quote(pathTransformFmt.substring(refIdx + 2))
+                : "";
+
+        transformed = prefix + matchPath + suffix;
+      }
+    }
```

Comment:
Also, this type of string manipulation would be greatly simplified with regex matching -- e.g. 

```java
Matcher m = Pattern.compile("(?<prefix>.*)%s(?<suffix>.*)").matcher(pathTransformFmt);
if (m.matches()) {
    String prefix = m.group("prefix") == null ? "" : Pattern.quote(m.group("prefix"));
    String suffix = m.group("suffix") == null ? "" : Pattern.quote(m.group("suffix"));
    String pattern = String.format("^%s%s%s$", prefix, matchPath, suffix);
    return Pattern.compile(pattern);
} else {...}
```

Reply:
This is subjective - I personally tend to understand the procedural code better than regular expressions. Also string concatenation is much more performant than string formatting which IMHO should be used sparingly and is much better suited for number formatting than string concatenation, but again that is somewhat subjective (barring the performance aspect, which is real).

Reply:
Heavy +1 to use regular expressions if possible. 3 if/else statements makes understanding not so easy.
So, I'm unresolving comment but feel free if you strongly are not agree
BTW Please add some documentation for `private static Pattern extractPathFromFmt` to explain what are you doing.

> Also string concatenation is much more performant than string formatting 

`String.format("^%s%s%s$", prefix, matchPath, suffix);` vs `'^' + prefix + matchPath + suffix + '$'` : both variants are OK for me in this particilar case